Mathlib/RingTheory/IntegralClosure.lean

```diff
@@ -973,135 +819,102 @@ open Algebra
 
-theorem isIntegral_trans_aux (x : B) {p : A[X]} (pmonic : Monic p) (hp : aeval x p = 0) :
-    IsIntegral (adjoin R (↑(p.map <| algebraMap A B).frange : Set B)) x := by
-  generalize hS : (↑(p.map <| algebraMap A B).frange : Set B) = S
-  have coeffs_mem : ∀ i, (p.map <| algebraMap A B).coeff i ∈ adjoin R S := by
-    intro i
-    by_cases hi : (p.map <| algebraMap A B).coeff i = 0
-    · rw [hi]
-      exact Subalgebra.zero_mem _
-    rw [← hS]
-    exact subset_adjoin (coeff_mem_frange _ _ hi)
-  obtain ⟨q, hq⟩ :
-    ∃ q : (adjoin R S)[X], q.map (algebraMap (adjoin R S) B) = (p.map <| algebraMap A B) := by
-    rw [← Set.mem_range]
-    exact (Polynomial.mem_map_range _).2 fun i => ⟨⟨_, coeffs_mem i⟩, rfl⟩
-  use q
-  constructor
-  · suffices h : (q.map (algebraMap (adjoin R S) B)).Monic
-    · refine' monic_of_injective _ h
-      exact Subtype.val_injective
-    · rw [hq]
-      exact pmonic.map _
-  · convert hp using 1
-    replace hq := congr_arg (eval x) hq
-    convert hq using 1 <;> symm <;> apply eval_map
-#align is_integral_trans_aux isIntegral_trans_aux
-
 variable [Algebra R A] [IsScalarTower R A B]
 
 /-- If A is an R-algebra all of whose elements are integral over R,
 and x is an element of an A-algebra that is integral over A, then x is integral over R.-/
 theorem isIntegral_trans (A_int : Algebra.IsIntegral R A) (x : B) (hx : IsIntegral A x) :
     IsIntegral R x := by
   rcases hx with ⟨p, pmonic, hp⟩
-  let S : Set B := ↑(p.map <| algebraMap A B).frange
-  refine' isIntegral_of_mem_of_FG (adjoin R (S ∪ {x})) _ _ (subset_adjoin <| Or.inr rfl)
-  refine' fg_trans (FG_adjoin_of_finite (Finset.finite_toSet _) fun x hx => _) _
-  · rw [Finset.mem_coe, frange, Finset.mem_image] at hx
-    rcases hx with ⟨i, _, rfl⟩
-    rw [coeff_map]
-    exact map_isIntegral (IsScalarTower.toAlgHom R A B) (A_int _)
-  · apply FG_adjoin_singleton_of_integral
-    exact isIntegral_trans_aux _ pmonic hp
+  let S := adjoin R (p.frange : Set A)
+  have : Module.Finite R S := ⟨(Subalgebra.toSubmodule S).fg_top.mpr <|
+    fg_adjoin_of_finite p.frange.finite_toSet fun a _ ↦ A_int a⟩
+  let p' : S[X] := p.toSubring S.toSubring fun a ha ↦ subset_adjoin ha
+  have hSx : IsIntegral S x := ⟨p', (p.monic_toSubring _ _).mpr pmonic, by
+    rw [IsScalarTower.algebraMap_eq S A B, ← eval₂_map]
+    convert hp; apply p.map_toSubring S.toSubring⟩
+  let Sx := Subalgebra.toSubmodule (adjoin S {x})
+  let MSx : Module S Sx := SMulMemClass.toModule _ -- the next line times out without this
+  have : Module.Finite S Sx := ⟨(Submodule.fg_top _).mpr hSx.fg_adjoin_singleton⟩
+  refine .of_mem_of_fg ((adjoin S {x}).restrictScalars R) ?_ _
+    ((Subalgebra.mem_restrictScalars R).mpr <| subset_adjoin rfl)
+  rw [← Submodule.fg_top, ← Module.finite_def]
+  letI : SMul S Sx := { MSx with } -- need this even though MSx is there
+  have : IsScalarTower R S Sx :=
+    Submodule.isScalarTower Sx -- Lean looks for `Module A Sx` without this
```

Comment:
Some typeclass inference failures that may be worth investigating (but are probably well known)